File: Beverages Storeroom App/app/src/main/java/com/afterten/drinks_transfers/data/SessionStore.kt
package com.afterten.drinks_transfers.data

import android.content.Context
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first

private val Context.sessionDataStore by preferencesDataStore(name = "session_store")

class SessionStore(private val context: Context) {
  private val tokenKey = stringPreferencesKey("token")
  private val userIdKey = stringPreferencesKey("user_id")
  private val emailKey = stringPreferencesKey("email")
  private val displayNameKey = stringPreferencesKey("display_name")
  private val loginAtKey = longPreferencesKey("login_at_ms")
  private val lastTransferWarehouseKey = stringPreferencesKey("last_transfer_to_warehouse")
  private val lastPurchaseSupplierKey = stringPreferencesKey("last_purchase_supplier")
  private val lastPurchaseWarehouseKey = stringPreferencesKey("last_purchase_warehouse")

  suspend fun readSession(): StoredSession? {
    val data = context.sessionDataStore.data.first()
    val token = data[tokenKey]
    val userId = data[userIdKey]
    val email = data[emailKey]
    val displayName = data[displayNameKey]
    val loginAt = data[loginAtKey]
    return if (!token.isNullOrBlank() && !userId.isNullOrBlank() && !email.isNullOrBlank() && loginAt != null) {
      StoredSession(token, userId, email, displayName, loginAt)
    } else {
      null
    }
  }

  suspend fun saveSession(
    token: String,
    userId: String,
    email: String,
    displayName: String?,
    loginAtMs: Long
  ) {
    context.sessionDataStore.edit { prefs ->
      prefs[tokenKey] = token
      prefs[userIdKey] = userId
      prefs[emailKey] = email
      if (displayName.isNullOrBlank()) {
        prefs.remove(displayNameKey)
      } else {
        prefs[displayNameKey] = displayName
      }
      prefs[loginAtKey] = loginAtMs
    }
  }

  suspend fun clearSession() {
    context.sessionDataStore.edit { prefs ->
      prefs.remove(tokenKey)
      prefs.remove(userIdKey)
      prefs.remove(emailKey)
      prefs.remove(displayNameKey)
      prefs.remove(loginAtKey)
    }
  }

  suspend fun getLastTransferWarehouseId(): String? {
    return context.sessionDataStore.data.first()[lastTransferWarehouseKey]
  }

  suspend fun setLastTransferWarehouseId(warehouseId: String) {
    context.sessionDataStore.edit { prefs ->
      prefs[lastTransferWarehouseKey] = warehouseId
    }
  }

  suspend fun getLastPurchaseSupplierId(): String? {
    return context.sessionDataStore.data.first()[lastPurchaseSupplierKey]
  }

  suspend fun setLastPurchaseSupplierId(supplierId: String) {
    context.sessionDataStore.edit { prefs ->
      prefs[lastPurchaseSupplierKey] = supplierId
    }
  }

  suspend fun getLastPurchaseWarehouseId(): String? {
    return context.sessionDataStore.data.first()[lastPurchaseWarehouseKey]
  }

  suspend fun setLastPurchaseWarehouseId(warehouseId: String) {
    context.sessionDataStore.edit { prefs ->
      prefs[lastPurchaseWarehouseKey] = warehouseId
    }
  }
}

data class StoredSession(
  val token: String,
  val userId: String,
  val email: String,
  val displayName: String?,
  val loginAtMs: Long
)

File: Beverages Storeroom App IOS/iosApp/Source/Services/SessionStore.swift
import Foundation

struct StoredSession {
  let token: String
  let userId: String
  let email: String
  let displayName: String?
  let loginAtMs: Int64
}

final class SessionStore {
  private let defaults = UserDefaults.standard

  private let tokenKey = "token"
  private let userIdKey = "user_id"
  private let emailKey = "email"
  private let displayNameKey = "display_name"
  private let loginAtKey = "login_at_ms"

  func readSession() -> StoredSession? {
    guard let token = defaults.string(forKey: tokenKey),
          let userId = defaults.string(forKey: userIdKey),
          let email = defaults.string(forKey: emailKey) else {
      return nil
    }
    let displayName = defaults.string(forKey: displayNameKey)
    let loginAt = defaults.object(forKey: loginAtKey) as? Int64
    if let loginAt = loginAt {
      return StoredSession(token: token, userId: userId, email: email, displayName: displayName, loginAtMs: loginAt)
    }
    return nil
  }

  func saveSession(token: String, userId: String, email: String, displayName: String?, loginAtMs: Int64) {
    defaults.set(token, forKey: tokenKey)
    defaults.set(userId, forKey: userIdKey)
    defaults.set(email, forKey: emailKey)
    defaults.set(displayName, forKey: displayNameKey)
    defaults.set(loginAtMs, forKey: loginAtKey)
  }

  func clearSession() {
    defaults.removeObject(forKey: tokenKey)
    defaults.removeObject(forKey: userIdKey)
    defaults.removeObject(forKey: emailKey)
    defaults.removeObject(forKey: displayNameKey)
    defaults.removeObject(forKey: loginAtKey)
  }
}
